### ISSUE TITLE
ubernetes: How to collect cluster resource metrics

### ISSUE THREAD

**Issue by nikhiljindal** [kubernetes/kubernetes#23614]
Forked from https://github.com/kubernetes/kubernetes/pull/23430#issuecomment-202768265

The ubernetes scheduler needs to know the available resources in each cluster to be able to divide resources (such as replicasets) appropriately.

As per https://github.com/kubernetes/kubernetes/blob/master/docs/design/federation-phase-1.md#cluster:

> In phase one it only contains available CPU resources and memory resources. The cluster controller will periodically poll the underlying cluster API Server to get cluster capability. In phase one it gets the metrics by simply aggregating metrics from all nodes. In future we will improve this with more efficient ways like leveraging heapster, and also more metrics will be supported.

There is also a proposal to provide cluster level metrics via apiserver: #23376, which we can start using whenever it is available.

cc @kubernetes/sig-cluster-federation

**Comment by nikhiljindal** [comment]
Copying @huangyuqi's comment from https://github.com/kubernetes/kubernetes/issues/23376#issuecomment-203185116:

> Hi, @vishh , I am working on Ubernetes recently.
> In ubernetes' design, metrics(CPU/MEM) in cluster level need to be collected, and used to make scheduling decisions.
> In my experience, the heapster has `Cluster-level Metrics`, like this:
> 
> ```
> /api/v1/model/metrics/: Returns a list of available cluster-level metrics.
> /api/v1/model/metrics/{metric-name}?start=X&end=Y: Returns a set of (Timestamp, Value) 
> pairs for the requested cluster-level metric, between the time range specified by start and end
> ```
> 
> So, @nikhiljindal , In my opinion , current heaspter can match our require about the cluster metrics. How do you think?

**Comment by nikhiljindal** [comment]
@huangyuqi I have no idea if heapster is currently setup to give us the metrics we want. Its great if it is :)

@vishh I think is the best person to answer that.

**Comment by huangyuqi** [comment]
@nikhiljindal , I am very familiar with @vishh and `heaspter`. 
In heapster, aggregation about the cluster level metrics will be taken. I will check it can match us completely or not.

**Comment by davidopp** [comment]
Please do not re-implement this functionality. If Heapster does not address your problem, then we will fix Heapster. We want to use a single metrics pipeline for all of the internal scheduling-related use cases. Ubernetes is one of these.

cc/ @fgrzadkowski @piosz

**Comment by mwielgus** [comment]
I agree with @davidopp. Heapster already aggregates metrics at the cluster level.

**Comment by vishh** [comment]
Given that Kubernetes itself doesn't support usage based scheduling, why does the ubernetes scheduler need this feature right away?
It might make sense to make it work in Kubernetes before attempting in Ubernetes.

**Comment by nikhiljindal** [comment]
Yes we might not need it right away.
In phase 1, we can simply schedule on any cluster assuming that there are sufficient resources.

But eventually, we want to ensure that the cluster we are scheduling on has sufficient resources.

**Comment by fgrzadkowski** [comment]
@nikhiljindal Is this issue about using usage metrics or amount of requested resources in a cluster?

Resource metrics API will be delivered for 1.3 so I'd strongly suggest to wait for it. Current plan is to expose resource usage metrics aggregated on pod and node level. I assume we can add cluster level aggregation. We have all of this data already in Heapster. I agree with @davidopp - let's make sure we don't reimplement metrics collection!

We should discuss this next week in MTV.

BTW. Heapster is now owned and maintained by @mwielgus and @piosz. Please CC them :)

**Comment by nikhiljindal** [comment]
Sure, thanks!

Its about computing the available resources in a cluster. Just the CPU and memory to start with.

**Comment by fgrzadkowski** [comment]
@nikhiljindal How do you define "available resource"? Do you mean CPU/memory usage or amount of requested resources (as is request/limit in PodSpec)?

**Comment by quinton-hoole** [comment]
The high-level goal here is for the Ubernetes scheduler to be able to rank clusters w.r.t available resources in order place replicas into clusters where they will both be able to be placed onto nodes ("scheduled"), and and have maximum access to required resources on those nodes.

I actually think that this is a hard problem, and should be deferred beyond phase 1.  My reasoning is as follows:

1. Knowing the aggregate amount of available (unreserved) resources in a cluster does not in fact enable the (Ubernetes) scheduler to accurately predict whether a pod/replica will successfully schedule onto a node in that cluster (e.g. 2GB of available RAM fragmented across 10 nodes does not necessarily allow a pod requiring, say 1GB to actually schedule).
2. It also does not enable the (Ubernetes) scheduler to predict how much resource will likely be available to a given replica even if it does get successfully scheduled onto a node (because other pods on a node might consume large amounts of CPU and RAM without explicitly reserving it, leaving relatively little for the pod in question).

One solution to both of these problems is to tentatively pick a cluster for a replica, and then reschedule it to other clusters if it does not run well (my empirical measure) in the originally chosen cluster (e.g. if it does not schedule for extended periods of time, or is repeatedly OOM killed etc).  This seems simpler to implement, and more effective overall.

So for phase 1 I think a reasonable approach is to pick a cluster at random to start with, and (in phase 2) to move replicas to other clusters if they do not run successfully.

**Comment by vishh** [comment]
> It also does not enable the (Ubernetes) scheduler to predict how much resource will likely be available to a given replica even if it does get successfully scheduled onto a node (because other pods on a node might consume large amounts of CPU and RAM without explicitly reserving it, leaving relatively little for the pod in question).

Quality of Service should take effect once a Pod is running in one of the nodes. These woes should be handled by QoS in general.

> One solution to both of these problems is to tentatively pick a cluster for a replica, and then reschedule it to other clusters if it does not run well (my empirical measure) in the originally chosen cluster (e.g. if it does not schedule for extended periods of time, or is repeatedly OOM killed etc). This seems simpler to implement, and more effective overall.

FYI: This is already the case. Kubelet can reject pods that the scheduler thought would fit on a node.

**Comment by mfanjie** [comment]
@mfanjie

**Comment by fgrzadkowski** [comment]
@quinton-hoole As @vishh explained QoS should do the job once we'll schedule the pod somewhere. Regular (non-ubernetes) scheduler also does not use usage metrics. I see no reason why it'd be more important in Ubernetes.

In other words I think it makes sense to take into account amount of requested resources (divided by capacity). This can be read directly from apiserver. Using usage metric is far more advanced and we don't do it in a regular cluster/scheduler, so we should not do it for Ubernetes.

**Comment by nikhiljindal** [comment]
Yes I agree with @quinton-hoole 
Lets start with choosing a cluster randomly.

We can add support for choosing a cluster more intelligently and moving resources to other clusters if they do not run successfully later.
Whenever we are ready to do that, the consensus seems to be that we should be using heapster to collect the stats for that.

thanks @vishh and @fgrzadkowski

**Comment by quinton-hoole** [comment]
@fgrzadkowski Yes, that makes perfect sense. As far as I know, amount of requested resources is not yet available via the API, but I'll check up on that.

On further reflection It also occurred to me that:

1. Although no scheme will be perfect w.r.t. predicting schedulability, we don't need to be perfect. All we need to do is rank clusters by probable schedulability.  I speculate that in the vast majority of cases, the cluster with the lowest percentage of reserved resources along the most contended resource axis (CPU, RAM etc) will be the on most likely to successfully schedule a pod, all other things being equal.  If the pod fails to schedule there in practise, we can either wait until it does, or try another cluster (or do both in parallel).

2. Yes, agreed, Q0S is sufficient to avoid resource starvation on already-scheduled pods, and if it's not, we should fix that in Kubernetes before Ubernetes.

**Comment by fgrzadkowski** [comment]
@quinton-hoole It's not exposed as an aggregated metric. Currently you'd need to manually add it after listing all pods. I think it makes perfect sense to add this metric to Resource Metric API @piosz

**Comment by quinton-hoole** [comment]
@fgrzadkowski @piosz Yes, I definitely don't want to be aggregating all nodes in Ubernetes.  Better done per cluster in Kubernetes.

**Comment by huangyuqi** [comment]
Hi, @nikhiljindal 
Choosing a cluster randomly you mentioned means that we don't need to get the status of kubernetes cluster, even if it is unhealthy, right? 
Or, the cluster list just contains healthy ones, and the cluster-controller gets the healthy condition by requesting kubernetes restapi `"/healthz"` and set the value.  If the response is not `"OK"`, the unhealthy cluster will remove from the cluster list which used to be selected randomly.

**Comment by irfanurrehman** [comment]
This issue was labelled only for sig/multicluster and is thus moved over to kubernetes/federation#41.
If this does not seem to be right, please reopen this and notify us @kubernetes/sig-multicluster-misc.
/close
